ionics_fits/models/quantum_phys.py

```diff
@@ -98,4 +98,42 @@ def squeezed_state_probs(
     return P_n
 
 
-# pytype: enable=invalid-annotation
+def thermal_displaced_state_probs(
+    n_max: int,
+    n_bar: ModelParameter(lower_bound=0, scale_func=scale_invariant),
+    alpha: ModelParameter(lower_bound=0, scale_func=scale_invariant),
+) -> Array[("num_fock_states",), np.float64]:
+    """Displaced thermal probability distribution.
+
+    For an ion initially in a thermal distribution characterised by an average phonon
+    number n_thermal, we calculate the new probability distribution after applying a
+    Displacement operator D(α).
+
+    formula taken from equation (7) of
+    Ramm, M., Pruttivarasin, T. and Häffner, H., 2014. Energy transport in trapped ion
+    chains. New Journal of Physics, 16(6), p.063062.
+    https://iopscience.iop.org/article/10.1088/1367-2630/16/6/063062/pdf
+
+    :param n_max: the distribution is truncated at a maximum Fock state of ``|n_max>``
+    :param n_bar: the mean Fock state occupation
+    :param alpha: Complex displacement parameter
+    :returns: array of Fock state occupation probabilities
+
+    """
+
+    n_alpha = alpha**2
+
+    n = np.arange(n_max + 1, dtype=int)
+    if n_bar == 0:
+        P_n = np.zeros_like(n)
+        P_n[0] = 1
+    else:
+        lag_poly = np.array(
+            [special.laguerre(n)(-n_alpha / (n_bar * (n_bar + 1))) for n in n]
+        )
+        P_n = (
+            (n_bar**n / (n_bar + 1) ** (n + 1))
```

Comment:
In `thermal_state_probs` we expressed this as `((n_bar / (n_bar + 1)) ** n) / (n_bar + 1)`. These are mathematically equivalent, but I wonder which is best?

My hunch is that the version in `thermal_state_probs` is probably the most accurate because we're taking the power of a number close to one. Can you have a think about numerical accuracy and speed here? Which ever way we choose to express this we should be consistent between the various functions.

Reply:
I looked at this for 10 minutes and saw no change in speed. In terms of numerical accuracy, the worse case difference was at <1e-20 level for the example I looked at - I would say this doesn't really matter. I just changed it to be the same format we use elsewhere.